l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,444 @@
+        bytes32 newHash = _hashValidatorPubKey(_pubKey);
+        validatorPubKeyHashes[newHash] = true;
+
+        _ensureValidatorSnapshot(node);
+        node.validatorLatest.pubKey = _pubKey;
+        node.validatorLatest.pop = _pop;
+
+        emit NodeValidatorKeyChanged(_nodeOwner, _pubKey, _pop);
+    }
+
+    /// @notice Changes the attester's public key of a node in the registry.
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner whose attester public key will be changed.
+    /// @param _pubKey The new ECDSA public key to assign to the node's attester.
+    function changeAttesterKey(
+        address _nodeOwner,
+        Secp256k1PublicKey calldata _pubKey
+    ) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        _verifyInputSecp256k1PublicKey(_pubKey);
+        _verifyNodeOwnerExists(_nodeOwner);
+        Node storage node = nodes[_nodeOwner];
+        if (_finalizeNodeRemoval(_nodeOwner, node)) {
+            return;
+        }
+
+        bytes32 prevHash = _hashAttesterPubKey(node.attesterLatest.pubKey);
+        delete attesterPubKeyHashes[prevHash];
+        bytes32 newHash = _hashAttesterPubKey(_pubKey);
+        attesterPubKeyHashes[newHash] = true;
+
+        _ensureAttesterSnapshot(node);
+        node.attesterLatest.pubKey = _pubKey;
+
+        emit NodeAttesterKeyChanged(_nodeOwner, _pubKey);
+    }
+
+    /// @notice Adds a new commit to the validator committee.
+    /// @dev Implicitly updates the validator committee by affecting off-chain readers based on the current state of a node's validator attributes:
+    /// - If "validatorsCommit" > "node.validatorLastUpdateCommit", read "node.validatorLatest".
+    /// - If "validatorsCommit" == "node.validatorLastUpdateCommit", read "node.validatorSnapshot".
+    /// @dev Only callable by the contract owner.
+    function commitValidators() external onlyOwner {
+        validatorsCommit++;
+
+        emit ValidatorsCommitted();
+    }
+
+    /// @notice Adds a new commit to the attester committee.
+    /// @dev Implicitly updates the attester committee by affecting off-chain readers based on the current state of a node's attester attributes:
+    /// - If "attestersCommit" > "node.attesterLastUpdateCommit", read "node.attesterLatest".
+    /// - If "attestersCommit" == "node.attesterLastUpdateCommit", read "node.attesterSnapshot".
+    /// @dev Only callable by the contract owner.
+    function commitAttesters() external onlyOwner {
+        attestersCommit++;
+
+        emit AttestersCommitted();
+    }
+
+    function numNodes() public view returns (uint256) {
+        return nodeOwners.length;
+    }
+
+    function _finalizeNodeRemoval(address _nodeOwner, Node storage _node) private returns (bool) {
+        if (
+            _node.attesterSnapshot.pendingRemoval &&
+            _node.validatorSnapshot.pendingRemoval
```

Comment:
Just wondering about whether we should consult `*Latest`s as well. 

As I understand the way e.g. `commitAttesters()` works is that it increments `attestersCommit`, and the reader of the state has to reason about `attesterLastUpdateCommit` to figure out if the committed state is `attesterSnapshot` or `attesterLatest`. Say we have the following: 
```
{ 
  attesterSnapshot: { pendingRemoval: false}, 
  attesterLatest: { pendingRemoval: true }, 
  attesterLastUpdateCommit: 10
}
```
and the current `attestersCommit` is 20, which mans `attesterLatest` is implicitly committed, it just happens that there has been no calls to any of the `change*` methods since commit 10. Then I can call e.g. `changeAttesterPublicKey` and it won't delete the record, instead it will update `attesterSnapshot.pendingRemoval` to `true`, and the _next_ time I call another `change*` method it will remove it (if I have also called changes on the validator side as well to update that snapshot).

Instead, `_finalizeNodeRemoval` could check `attesterLatestUpdateCommit` and `attesterCommit` like offline readers are supposed to, and delete the record without requiring an extra change.

Maybe this can be an accessor method on the struct?

Reply:
I agree, the current implementation introduces additional laziness, due to checking prior to doing the actual attribute change, and the shift onto snapshot. I had the improvement in the pipeline. 

Added here: https://github.com/matter-labs/era-contracts/pull/555/commits/655f4b5d9ce76e427384f6121c35cd8862f57285

I didn’t define the inspection function (`_isNodePendingRemoval(Node storage _node)`) as a method because the struct is defined in the interface, and it would turn out a bit cumbersome.